src/repo.rs

```diff
@@ -59,6 +59,42 @@ impl Repository {
         }
     }
 
+    /// Find and open an existing repository, with additional options.
+    ///
+    /// If flags contains REPOSITORY_OPEN_NO_SEARCH, the path must point
+    /// directly to a repository; otherwise, this may point to a subdirectory
+    /// of a repository, and `open_ext` will search up through parent
+    /// directories.
+    ///
+    /// If flags contains REPOSITORY_OPEN_CROSS_FS, the search through parent
+    /// directories will not cross a filesystem boundary (detected when the
+    /// stat st_dev field changes).
+    ///
+    /// If flags contains REPOSITORY_OPEN_BARE, force opening the repository as
+    /// bare even if it isn't, ignoring any working directory, and defer
+    /// loading the repository configuration for performance.
+    ///
+    /// ceiling_dirs specifies a list of paths (separated by the platform path
+    /// list separator) that the search through parent directories will stop
+    /// before entering.  Use the functions in std::env to construct or
+    /// manipulate such a path list.
+    pub fn open_ext<P: AsRef<Path>>(path: P,
+                                    flags: RepositoryOpenFlags,
+                                    ceiling_dirs: Option<&str>)
```

Comment:
You mentioned that this may be more convenient if the caller already has this format expected by libgit2, but my counter point to that would be:

* No one actually ever has that format expected by libgit2
* If you did, it'd likely be the return value of `env::join_paths` which is an `OsString`

For the most ergonomic usage this should probably take a `I: IntoIterator<T: IntoCString>`, but I'd be fine just taking a slice for now.